src/python/pants/backend/native/tasks/native_task.py

```diff
@@ -72,23 +73,32 @@ def implementation_version(cls):
   def _native_build_settings(self):
     return NativeBuildSettings.global_instance()
 
+  @memoized_property
+  def _native_build_step_settings(self):
+    return NativeBuildStepSettings.global_instance()
+
   @memoized_property
   def _native_toolchain(self):
     return NativeToolchain.scoped_instance(self)
 
-  @memoized_property
-  def _toolchain_variant_request(self):
+  def _toolchain_variant_request(self, variant):
     return ToolchainVariantRequest(
       toolchain=self._native_toolchain,
-      variant=self._native_build_settings.toolchain_variant)
+      variant=variant)
 
-  @memoized_method
-  def get_c_toolchain_variant(self):
-    return self._request_single(CToolchain, self._toolchain_variant_request)
+  def get_c_toolchain_variant(self, native_library_target):
+    return self._get_toolchain_variant(CToolchain, native_library_target)
 
-  @memoized_method
-  def get_cpp_toolchain_variant(self):
-    return self._request_single(CppToolchain, self._toolchain_variant_request)
+  def get_cpp_toolchain_variant(self, native_library_target):
+    return self._get_toolchain_variant(CppToolchain, native_library_target)
+
+  def _get_toolchain_variant(self, toolchain_type, native_library_target):
+    if native_library_target.toolchain_variant is not None:
+      return self._request_single(toolchain_type, self._toolchain_variant_request(
+        native_library_target.toolchain_variant))
+    else:
+      return self._request_single(toolchain_type, self._toolchain_variant_request(
+        self._native_build_step_settings.toolchain_variant))
```

Comment:
This is great, but also duplicates the logic in `MirroredTargetOptionMixin`, which makes it harder to implement #7183 in the future. Could you please create a method as described in another comment on `native_build_step_settings.py`, and turn this into:
```python
    selected_variant = self._native_build_step_settings.get_toolchain_variant_for_target(native_library_target)
    return self._request_single(toolchain_type, self._toolchain_variant_request(selected_variant))
```